sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/collectionOperations.scala

```diff
@@ -2355,3 +2355,347 @@ case class ArrayRemove(left: Expression, right: Expression)
+            while (i < array.numElements()) {
+              if (hsSize.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+                ArraySetLike.throwUnionLengthOverflowException(hsSize.size)
+              }
+              hsSize.add(array.getInt(i))
+              i += 1
+            }
+          })
+          if (UnsafeArrayData.useGenericArrayData(IntegerType.defaultSize, hsSize.size)) {
+            ArrayUnion.evalUnionContainsNull(array1, array2, elementType,
+              if (elementTypeSupportEquals) null else ordering)
+          } else {
+            hsInt = new OpenHashSet[Int]
+            val resultArray = UnsafeArrayData.fromPrimitiveArray(new Array[Int](hsSize.size))
+            evalPrimitiveType(array1, array2, hsSize.size, resultArray, false)
+          }
+        case LongType =>
+          // avoid boxing of primitive long array elements
+          // calculate result array size
+          val hsSize = new OpenHashSet[Long]
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements()) {
+              if (hsSize.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+                ArraySetLike.throwUnionLengthOverflowException(hsSize.size)
+              }
+              hsSize.add(array.getLong(i))
+              i += 1
+            }
+          })
+          if (UnsafeArrayData.useGenericArrayData(IntegerType.defaultSize, hsSize.size)) {
+            ArrayUnion.evalUnionContainsNull(array1, array2, elementType,
+              if (elementTypeSupportEquals) null else ordering)
+          } else {
+            hsLong = new OpenHashSet[Long]
+            val resultArray = UnsafeArrayData.fromPrimitiveArray(new Array[Long](hsSize.size))
+            evalPrimitiveType(array1, array2, hsSize.size, resultArray, true)
+          }
+        case _ =>
+          val arrayBuffer = new scala.collection.mutable.ArrayBuffer[Any]
+          val hs = new OpenHashSet[Any]
+          var foundNullElement = false
+          Seq(array1, array2).foreach(array => {
+            var i = 0
+            while (i < array.numElements()) {
+              if (array.isNullAt(i)) {
+                if (!foundNullElement) {
+                  arrayBuffer += null
+                  foundNullElement = true
+                }
+              } else {
+                val elem = array.get(i, elementType)
+                if (!hs.contains(elem)) {
+                  if (arrayBuffer.size > ByteArrayMethods.MAX_ROUNDED_ARRAY_LENGTH) {
+                    ArraySetLike.throwUnionLengthOverflowException(arrayBuffer.size)
+                  }
+                  arrayBuffer += elem
+                  hs.add(elem)
+                }
+              }
+              i += 1
+            }
+          })
+          new GenericArrayData(arrayBuffer)
+      }
+    } else {
+      ArrayUnion.evalUnionContainsNull(array1, array2, elementType,
+        if (elementTypeSupportEquals) null else ordering)
```

Comment:
Now that we don't need to check `elementTypeSupportEquals` and can always use `ordering`?

Reply:
From the view of correctness, it work only with `ordering`.

If we use `ordering for elementTypeSupportEquals`, we have to use a loop instead of a hash function in `evalUnionContainsNull`. I think that to use a loop will lead to performance degradation compared to use a `mutable.HashSet`.
This is why we try not to use `ordering` as possible. WDYT?

Reply:
nvm, got it. Now, here `elementTypeSupportEquals ` is always `false`.